### PR TITLE
feat: (levm) arithmetic operations

### LINKED ISSUE #457: Opcode: ADD

### LINKED ISSUE #458: Opcode: MUL

### LINKED ISSUE #459: Opcode: SUB

### LINKED ISSUE #460: Opcode: DIV

### LINKED ISSUE #461: Opcode: SDIV

### LINKED ISSUE #462: Opcode: MOD

### LINKED ISSUE #463: Opcode: SMOD

### LINKED ISSUE #464: Opcode: ADDMOD

### LINKED ISSUE #465: Opcode: MULMOD

### LINKED ISSUE #466: Opcode: EXP

### LINKED ISSUE #467: Opcode: SIGNEXTEND

### PR DESCRIPTION
**Pull request by belfortep**
**Motivation**

This PR adds the arithmetic operations for LEVM

**Description**

Adds the implementation of this group of opcodes with tests

Closes #457 
Closes #458
Closes #459 
Closes #460 
Closes #461
Closes #462
Closes #463 
Closes #464
Closes #465 
Closes #466 
Closes #467

### CODE REVIEW

File: crates/levm/src/vm.rs [hunk 1]
```diff
@@ -9,6 +9,15 @@ pub struct VM {
     pub pc: usize,
 }
 
+/// Shifts the value to the right by 255 bits and checks the most significant bit is a 1
+fn is_negative(value: U256) -> bool {
+    value >> 255 == U256::one()
+}
+/// Converts a positive value to a negative one in two's complement
+fn to_negative(value: U256) -> U256 {
```
Comment (MegaRedHand) [inline crates/levm/src/vm.rs:17]: Can we rename this to `negate` or similar? Since it also turns negative numbers into positive ones.

File: crates/levm/src/vm.rs [hunk 2]
```diff
@@ -17,7 +26,127 @@ impl VM {
...
+                    let bits_per_byte = U256::from(8);
+                    let sign_bit_position_on_byte = 7;
+                    let max_byte_size = 31;
+
+                    let byte_size = byte_size.min(U256::from(max_byte_size));
+                    let sign_bit_index = bits_per_byte * byte_size + sign_bit_position_on_byte;
+                    let is_negative = value_to_extend.bit(sign_bit_index.as_usize());
+                    let sign_bit_mask = (U256::one() << sign_bit_index) - U256::one();
+                    let result = if is_negative {
+                        value_to_extend.saturating_add(!sign_bit_mask)
```
Comment (MegaRedHand) [inline crates/levm/src/vm.rs:145]: I think we want to apply bitwise OR `value_to_extend | !sign_bit_mask` here.

File: crates/levm/tests/tests.rs [hunk 3]
```diff
@@ -21,9 +23,394 @@ fn test() {
...
+        .flat_map(Operation::to_bytecode)
+        .collect::<Bytes>();
+
+    vm.execute(bytecode);
+
+    assert!(vm.stack.pop().unwrap() == U256::from(4));
+}
+
+#[test]
+fn addmod_op_for_zero() {
```
Comment (MegaRedHand) [inline crates/levm/tests/tests.rs:288]: Can you check that `ADDMOD` works correctly when adding `2^256 - 2` and `2^256 - 3` modulo `2^256 - 1`?

File: crates/levm/tests/tests.rs [hunk 4]
```diff
@@ -21,9 +23,394 @@ fn test() {
...
+        .flat_map(Operation::to_bytecode)
+        .collect::<Bytes>();
+
+    vm.execute(bytecode);
+
+    assert!(vm.stack.pop().unwrap() == U256::zero());
+}
+
+#[test]
+fn mulmod_op() {
```
Comment (MegaRedHand) [inline crates/levm/tests/tests.rs:310]: Let's test this with big numbers that trigger the overflowing behaviour, and modules comparable to 2^256

**Review by iinaki — APPROVED** [review]
nice work!

**Review by ilitteri — COMMENTED** [review]
The PR looks good and I agree with @MegaRedHand comments. I'd also add a change in the naming. Let's use the formal operand names for the operations (both in implementation and tests if needed), which are:

- Addition: `augend + addend = sum`.
- Subtraction: `minuend - subtrahend = difference`.
- Multiplication: `multiplicand × multiplier = product`.
- Division: `dividend ÷ divisor = quotient`.
- Modulo: `dividend % divisor = remainder`.
- Exponentiation: `base ^ exponent = power`.

File: crates/levm/tests/tests.rs (hunk 4 above)
Comment (belfortep) [reply crates/levm/tests/tests.rs:310]: Should we add this tests for all the opcodes that could overflow?

File: crates/levm/tests/tests.rs (hunk 3 above)
Comment (belfortep) [reply crates/levm/tests/tests.rs:288]: Checked it and the code had a bug returning a wrong value, just fixed it and added a test 🫡

File: crates/levm/tests/tests.rs (hunk 4 above)
Comment (MegaRedHand) [reply crates/levm/tests/tests.rs:364]: Sure!

**Review by ilitteri — APPROVED** [review]
LGTM
